darshan-runtime/lib/darshan-pnetcdf-api.m4

```diff
@@ -0,0 +1,760 @@
+                tm1, tm2, rec_ref->last_meta_end);
+        }
+        PNETCDF_FILE_POST_RECORD();
+    }
+    return(ret);
+}
+
+DARSHAN_FORWARD_DECL(ncmpi_sync, int, (int ncid));
+
+int DARSHAN_DECL(ncmpi_sync)(int ncid)
+{
+    int ret;
+    double tm1, tm2;
+
+    MAP_OR_FAIL(ncmpi_sync);
+
+    tm1 = PNETCDF_WTIME();
+    ret = __real_ncmpi_sync(ncid);
+    tm2 = PNETCDF_WTIME();
+
+    if (ret == NC_NOERR) {
+        PNETCDF_FILE_PRE_RECORD();
+        struct pnetcdf_file_record_ref *rec_ref;
+        rec_ref = darshan_lookup_record_ref(pnetcdf_file_runtime->ncid_hash, &ncid, sizeof(int));
+        if (rec_ref)
+        {
+            rec_ref->file_rec->counters[PNETCDF_FILE_SYNCS] += 1;
+            DARSHAN_TIMER_INC_NO_OVERLAP(
+                rec_ref->file_rec->fcounters[PNETCDF_FILE_F_META_TIME],
+                tm1, tm2, rec_ref->last_meta_end);
+        }
+        PNETCDF_FILE_POST_RECORD();
+    }
+    return(ret);
+}
+
+dnl
+dnl PNETCDF_VAR_RECORD_OPEN(ncid, name, xtype, ndims, dimids, varidp, tm1, tm2)
+dnl
+define(`PNETCDF_VAR_RECORD_OPEN',`
+    do {
+        char *file_path, *rec_name;
+        int err, i, ret_name_len, unlimdimid, type_size;
+        darshan_record_id rec_id;
+        struct pnetcdf_var_record_ref *rec_ref;
+        MPI_Offset npoints;
+
+        /* get corresponding file name */
+        err = ncmpi_inq_path($1, &ret_name_len, NULL);
+        if (err != NC_NOERR) break;
+        rec_name = (char*) malloc(ret_name_len+1);
+        err = ncmpi_inq_path($1, NULL, rec_name);
+        if (err != NC_NOERR) { free(rec_name); break; }
+        /* fully resolve file path */
+        file_path = darshan_clean_file_path(rec_name);
+        free(rec_name);
+        rec_name = (char*) malloc(strlen(file_path)+strlen($2)+2);
+        strcpy(rec_name, file_path);
+        free(file_path);
+        /* append variable name if we have space */
+        strcat(rec_name, DARSHAN_PNETCDF_DATASET_DELIM);
+        strcat(rec_name, $2);
+        rec_id = darshan_core_gen_record_id(rec_name);
+        rec_ref = darshan_lookup_record_ref(pnetcdf_var_runtime->rec_id_hash, &rec_id, sizeof(darshan_record_id));
+        if (!rec_ref) rec_ref = pnetcdf_var_track_new_record(rec_id, rec_name);
+        free(rec_name);
+        if (!rec_ref) break;
+        rec_ref->var_rec->counters[PNETCDF_VAR_OPENS] += 1;
+        if (rec_ref->var_rec->fcounters[PNETCDF_VAR_F_OPEN_START_TIMESTAMP] == 0 ||
+            rec_ref->var_rec->fcounters[PNETCDF_VAR_F_OPEN_START_TIMESTAMP] > $7)
+            rec_ref->var_rec->fcounters[PNETCDF_VAR_F_OPEN_START_TIMESTAMP] = $7;
+        rec_ref->var_rec->fcounters[PNETCDF_VAR_F_OPEN_END_TIMESTAMP] = $8;
+        DARSHAN_TIMER_INC_NO_OVERLAP(rec_ref->var_rec->fcounters[PNETCDF_VAR_F_META_TIME],
+            $7, $8, rec_ref->last_meta_end);
+        err = ncmpi_inq_unlimdim($1, &unlimdimid);
+        i = ($5[0] == unlimdimid) ? 1 : 0; /* record variable or not */
```

Comment:
I am not sure I understand the question.
Is there a problem for npoints always be > 0?
unlimited dimension in PnetCDF and HDF5 means the variable size can
be extended along that dimension. The size of unlimited dimension will
be a number indicating its current size.

Reply:
I'm mostly just wondering whether it's useful to know whether a particular variable has an unlimited dimension using Darshan counters. If so, we should make sure we capture it somehow, because currently there's no way of knowing.

I see that PnetCDF has the ability to return the current size of the unlimited dimension, but this code block intentionally skips the unlimited dimension when calculating `npoints`. Should it be updated to include the current size of the unlimited dimension?

Reply:
The value i in indicates whether or not this variable has an unlimited dimension.
```
        i = ($5[0] == unlimdimid) ? 1 : 0; /* record variable or not */
```
The size of such variable cannot be known at the time it is defined, because
it can grow as applications write more data which extends that dimension size.

Does Darshan collect the dimension sizes of all variables?

Reply:
The PnetCDF module is attempting to set both the NDIMS and NPOINTS for each accessed variable. I guess for a variable with an unlimited dimension, there is no way to provide an exact NPOINTS at variable definition time, so maybe the best you can do is the NPOINTS in all other dimensions.

If we do not care about being able to detect whether a variable has an unlimited dimension via a Darshan log, then I guess this is fine. That's really my main point. If there is any use at all in knowing whether a PnetCDF variable can grow, then we need to add an additional counter.

Reply:
NPOINTS is only set once at variable definition time, it is not something that is currently updated on every I/O call. Given that, the question is what it should be set to for variables with an unlimited dimension? What if I have a variable that has 1 dimension only, and it can grow unlimited? Currently, NPOINTS would be set to 1 for this type of variable. We can also get an NPOINTS of 1 for a scalar variable, though, so I find that misleading that a scalar and an unlimited array would both report NPOINTS == 1.

Reply:
NPOINTS is calculated in two places: ncmpi_inq_varid and ncmpi_def_var.
So, when a put/get API is called, NPOINTS has been calculated.

In PnetCDF, a variable can be defined with the unlimited dimension and
never been written. In this case, the size of the variable is zero. Make sense?